cpu/stm32_common/periph/usbdev.c

```diff
@@ -0,0 +1,1168 @@
+            }
+        }
+    }
+    return ep;
+}
+
+/**
+ * @brief reset a TX fifo.
+ *
+ * @param   conf      usbdev context
+ * @param   fifo_num    fifo number to reset, 0x10 for all fifos
+ */
+static void _flush_fifo(const stm32_usb_otg_fshs_config_t *conf,
+                        uint8_t fifo_num)
+{
+    uint32_t reg = _global_regs(conf)->GRSTCTL & ~(USB_OTG_GRSTCTL_TXFNUM);
+
+    reg |= fifo_num << USB_OTG_GRSTCTL_TXFNUM_Pos | USB_OTG_GRSTCTL_TXFFLSH;
+    _global_regs(conf)->GRSTCTL = reg;
+    /* Wait for flush to finish */
+    while (_global_regs(conf)->GRSTCTL & USB_OTG_GRSTCTL_TXFFLSH) {}
+}
+
+static void _flush_rx_fifo(const stm32_usb_otg_fshs_config_t *conf)
+{
+    _global_regs(conf)->GRSTCTL |= USB_OTG_GRSTCTL_RXFFLSH;
+    while (_global_regs(conf)->GRSTCTL & USB_OTG_GRSTCTL_RXFFLSH) {}
+}
+
+static void _sleep_periph(const stm32_usb_otg_fshs_config_t *conf)
+{
+    *_pcgcctl_reg(conf) |= USB_OTG_PCGCCTL_STOPCLK;
+    /* Unblocking STM32_PM_STOP during suspend on the stm32f446 breaks
+     * while (un)blocking works on the stm32f401, needs more
+     * investigation with a larger set of chips */
+#ifdef STM32_USB_OTG_CID_1x
+    pm_unblock(STM32_PM_STOP);
+#endif
+}
+
+static void _wake_periph(const stm32_usb_otg_fshs_config_t *conf)
+{
+#ifdef STM32_USB_OTG_CID_1x
+    pm_block(STM32_PM_STOP);
+#endif
+    *_pcgcctl_reg(conf) &= ~USB_OTG_PCGCCTL_STOPCLK;
+    _flush_rx_fifo(conf);
+    _flush_fifo(conf, 0x10);
+}
+
+static void _reset_eps(stm32_usb_otg_fshs_t *usbdev)
+{
+    const stm32_usb_otg_fshs_config_t *conf = usbdev->config;
+
+    /* Set the NAK for all endpoints */
+    for (size_t i = 0; i < _max_endpoints(conf); i++) {
+        _out_regs(conf, i)->DOEPCTL |= USB_OTG_DOEPCTL_SNAK;
+        _in_regs(conf, i)->DIEPCTL |= USB_OTG_DIEPCTL_SNAK;
+        _in_regs(conf, i)->DIEPCTL |= (i) << USB_OTG_DIEPCTL_TXFNUM_Pos;
+    }
+}
+
+static void _reset_periph(stm32_usb_otg_fshs_t *usbdev)
+{
+    const stm32_usb_otg_fshs_config_t *conf = usbdev->config;
+
+    /* Wait for AHB idle */
+    while (!(_global_regs(conf)->GRSTCTL & USB_OTG_GRSTCTL_AHBIDL)) {}
+    _global_regs(conf)->GRSTCTL |= USB_OTG_GRSTCTL_CSRST;
+    /* Wait for reset done */
+    while (_global_regs(conf)->GRSTCTL & USB_OTG_GRSTCTL_CSRST) {}
+}
+
+static void _enable_gpio(const stm32_usb_otg_fshs_config_t *conf)
+{
+    /* Enables clock on the GPIO bus */
+    gpio_init(conf->dp, GPIO_IN);
+    gpio_init(conf->dm, GPIO_IN);
+    /* Configure AF for the pins */
+    gpio_init_af(conf->dp, conf->af);
+    gpio_init_af(conf->dm, conf->af);
+}
+
+static void _set_mode_device(stm32_usb_otg_fshs_t *usbdev)
+{
+    const stm32_usb_otg_fshs_config_t *conf = usbdev->config;
+
+    /* Force device mode */
+    _global_regs(conf)->GUSBCFG |= USB_OTG_GUSBCFG_FDMOD;
+    /* Spinlock to prevent a context switch here, needs a delay of 25 ms when
+     * force switching mode */
+    xtimer_spin(xtimer_ticks_from_usec(25 * US_PER_MS));
```

Comment:
You can get rid of the xtimer dependency by simply use a while loop and the value of CLOCK_CORECLOCK to spin during 25ms.

Reply:
Only if your CPU does one loop iteration per clock cycle.